File: tech-studio-projects/financial-industry/algo-trading-platform/cpp/orderbook/OrderbookSpread.cpp
#include "OrderbookSpread.h"

// Constructor implementation.
OrderbookSpread::OrderbookSpread(std::optional<long> bid, std::optional<long> ask)
    : Bid(bid), Ask(ask) {
    // Initialization is handled in the member initializer list.
}

// Getter for the bid price.
std::optional<long> OrderbookSpread::GetBid() const {
    return Bid;
}

// Getter for the ask price.
std::optional<long> OrderbookSpread::GetAsk() const {
    return Ask;
}

// Getter for the spread between ask and bid prices.
std::optional<long> OrderbookSpread::GetSpread() const {
    if (Bid.has_value() && Ask.has_value()) {
        return Ask.value() - Bid.value();
    }
    return std::nullopt; // Return std::nullopt if either Bid or Ask is not present.
}

File: tech-studio-projects/financial-industry/algo-trading-platform/java/algo-trading-platform/src/com/fintech/algotrading/orderbook/OrderbookSpread.java
package com.fintech.algotrading.orderbook;

import java.util.Optional;

public class OrderbookSpread {

    private final Optional<Long> bid;
    private final Optional<Long> ask;

    // Constructor
    public OrderbookSpread(Optional<Long> bid, Optional<Long> ask) {
        this.bid = bid;
        this.ask = ask;
    }

    // Getter for the bid price
    public Optional<Long> getBid() {
        return bid;
    }

    // Getter for the ask price
    public Optional<Long> getAsk() {
        return ask;
    }

    // Getter for the spread between ask and bid prices
    public Optional<Long> getSpread() {
        if (bid.isPresent() && ask.isPresent()) {
            return Optional.of(ask.get() - bid.get());
        }
        return Optional.empty(); // Return Optional.empty() if either bid or ask is not present
    }
}

